v2/middlewares/retry/retry.go

```diff
@@ -0,0 +1,37 @@
+package retry
+
+import (
+	"context"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+
+	"github.com/honestbank/kp/v2/internal/retrycounter"
+	"github.com/honestbank/kp/v2/middlewares"
+)
+
+type retry struct {
+	producer        Producer
+	onProduceErrors func(err error)
+}
+
+type Producer interface {
+	ProduceRaw(item *kafka.Message) error
+}
+
+func (r retry) Process(ctx context.Context, item *kafka.Message, next func(ctx context.Context, item *kafka.Message) error) error {
+	err := next(ctx, item)
+	if err == nil {
+		return nil
+	}
+	retrycounter.SetCount(item, retrycounter.GetCount(item)+1)
+	err = r.producer.ProduceRaw(&kafka.Message{Value: item.Value, Key: item.Key, Headers: item.Headers, Timestamp: item.Timestamp, TimestampType: item.TimestampType, Opaque: item.Opaque})
```

Comment:
probably worth checking the retrycounter here as well.

This code works well when user do
```
processor.
AddMiddleware(retryMiddleware).
AddMiddleware(deadLetterMiddleware)
```
but we cannot assume it's always in this order.

Reply:
as discussed, I'll document order of middleware being important in the documentation on #46